### PR TITLE
Additional tests

### PR DESCRIPTION
**Pull request by brianpopow**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

This PR adds additional unit tests for:

- Setting iptc value with strict option disabled
- Tests for jpg encoder preserving exif, iptc, icc metadata when encoding a image
- Additional test cases for png: Some edge cases and some additional png variants. Note: I have splitted up the `Common` test cases in tests with some more meaningful names.

Another change was made for #410: When a chunk is detected which is from a apple PNG which does not follow the spec, a exception with a better error message is thrown.

I noticed that we do not check the CRC for the IDAT chunks of a PNG. I could not figure out a good solution for that without reading the IDAT chunk twice.

### CODE REVIEW

File: src/ImageSharp/Formats/Png/PngChunkType.cs [hunk 1]
```diff
@@ -73,6 +73,58 @@ internal enum PngChunkType : uint
...
+        /// The cHRM chunk may be used to specify the 1931 CIE x,y chromaticities of the red,
+        /// green, and blue display primaries used in the image, and the referenced white point.
+        /// </summary>
+        Chroma = 0x6348524d,
+
+        /// <summary>
+        /// Malformed chunk named CgBI produced by apple, which is not conform to the specification.
+        /// Related issue is here https://github.com/SixLabors/ImageSharp/issues/410
+        /// </summary>
+        MalformedApple = 0x43674249
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngChunkType.cs:128]: `ProprietaryApple`. I think we could theoretically support this in the future.

- Skip Zlib header, footer, and CRC when reading the IDAT chunk
- Read as premultiplied Bgra instead of Rgba

http://iphonedevwiki.net/index.php/CgBI_file_format#CgBI_Chunk

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs [hunk 2]
```diff
@@ -215,6 +215,9 @@ public Image<TPixel> Decode<TPixel>(Stream stream)
                             case PngChunkType.End:
                                 this.isEndChunkReached = true;
                                 break;
+                            case PngChunkType.MalformedApple:
+                                PngThrowHelper.ThrowInvalidChunkType("Malformed Apple PNG detected! This PNG file is not conform to the specification and cannot be decoded.");
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:219]:
```suggestion
                                PngThrowHelper.ThrowInvalidChunkType("Proprietary Apple PNG detected! This PNG file is not conform to the specification and cannot be decoded.");
```

File: src/ImageSharp/Formats/Png/PngChunkType.cs (hunk 1 above)
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngChunkType.cs:128]:
```suggestion
        ProprietaryApple = 0x43674249
```

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs [hunk 3]
```diff
@@ -215,6 +215,9 @@ public Image<TPixel> Decode<TPixel>(Stream stream)
                             case PngChunkType.End:
                                 this.isEndChunkReached = true;
                                 break;
+                            case PngChunkType.MalformedApple:
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:218]:
```suggestion
                            case PngChunkType.ProprietaryApple:
```

File: src/ImageSharp/Formats/Png/PngThrowHelper.cs [hunk 4]
```diff
@@ -20,11 +20,14 @@ internal static class PngThrowHelper
...
         public static void ThrowInvalidChunkType() => throw new InvalidImageContentException("Invalid PNG data.");
 
+        [MethodImpl(InliningOptions.ColdPath)]
+        public static void ThrowInvalidChunkType(string message) => throw new InvalidImageContentException(message);
+
         [MethodImpl(InliningOptions.ColdPath)]
         public static void ThrowInvalidChunkCrc(string chunkTypeName) => throw new InvalidImageContentException($"CRC Error. PNG {chunkTypeName} chunk is corrupt!");
 
         [MethodImpl(InliningOptions.ColdPath)]
-        public static void ThrowNotSupportedColor() => new NotSupportedException("Unsupported PNG color type");
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngThrowHelper.cs:27]: Good catch!

File: src/ImageSharp/Formats/Png/PngChunkType.cs (hunk 1 above)
Comment (brianpopow) [reply src/ImageSharp/Formats/Png/PngChunkType.cs]: yeah theoretical, but who knows what else from the spec they do not follow.

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngChunkType.cs]: Yeah, could be troublesome. Something for the future anyway....
